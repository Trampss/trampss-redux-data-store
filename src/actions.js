const scope = action => `@trampss/${action.toUpperCase()}`

export const SET = prefix => scope(`SET_${prefix}`)
export const set = prefix => payload => ({ type: SET(prefix), payload })

export const RESET = prefix => scope(`RESET_${prefix}`)
export const reset = prefix => () => ({ type: RESET(prefix) })

export const ADD = prefix => scope(`ADD_${prefix}`)
export const add = prefix => payload => ({ type: ADD(prefix), payload })

export const UPDATE = prefix => scope(`UPDATE_${prefix}`)
export const update = prefix => payload => ({ type: UPDATE(prefix), payload })

export const REMOVE = prefix => scope(`REMOVE_${prefix}`)
export const remove = prefix => key => ({ type: REMOVE(prefix), payload: key })

export const ADD_OR_UPDATE = prefix => scope(`ADD_OR_UPDATE_${prefix}`)
export const addOrUpdate = prefix => payload => ({ type: ADD_OR_UPDATE(prefix), payload })
