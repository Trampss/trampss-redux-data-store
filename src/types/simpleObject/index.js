import middleware from './simpleObject.middleware'

export default {
  middlewares: [middleware],
  actions: ['set', 'reset', 'update'],
  selectors: ['getState', 'isInitialized', 'get'],
}
