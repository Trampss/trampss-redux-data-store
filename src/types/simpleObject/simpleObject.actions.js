export { set, SET, reset, RESET, update, UPDATE } from '../../actions'
