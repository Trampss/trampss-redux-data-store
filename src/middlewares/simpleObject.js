import { SET, RESET, UPDATE } from '../actions'

export const initState = { data: undefined, initialized: false }

const init = defaultData => ({ ...initState, data: defaultData })

const reducer = (/* key */) => prefix => defaultData =>
  (state = init(defaultData), { type, payload } = {}) => {
    switch (type) {
      case SET(prefix):
        return {
          data: payload,
          initialized: true,
        }
      case UPDATE(prefix):
        return {
          ...state,
          data: {
            ...state.data,
            ...payload,
          },
        }
      case RESET(prefix):
        return init(defaultData)
      default:
        return state
    }
  }

export default key => prefix => defaultData => (ctx = {}) => ({
  ...ctx,
  state: reducer(key)(prefix)(defaultData)(ctx.state, ctx.action),
})
