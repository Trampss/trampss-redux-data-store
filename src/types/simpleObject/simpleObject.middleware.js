import { SET, RESET, UPDATE } from '../../actions'

export const initState = {}
const defaultState = defaultData => defaultData !== undefined ? defaultData : initState

const reducer = (/* key */) => prefix => defaultData =>
  (state = defaultState(defaultData), { type, payload } = {}) => {
    switch (type) {
      case SET(prefix):
        return payload
      case UPDATE(prefix):
        return {
          ...state,
          ...payload,
        }
      case RESET(prefix):
        return defaultState(defaultData)
      default:
        return state
    }
  }

export default key => prefix => defaultData => (ctx = {}) => ({
  ...ctx,
  state: reducer(key)(prefix)(defaultData)(ctx.state, ctx.action),
})
