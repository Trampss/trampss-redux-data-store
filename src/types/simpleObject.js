import simpleObject from '../middlewares/simpleObject'

export default {
  middlewares: [simpleObject],
  actions: ['set', 'reset'],
  selectors: ['getState', 'isInitialized', 'get'],
}
