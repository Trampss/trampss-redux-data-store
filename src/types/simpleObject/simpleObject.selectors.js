/* eslint-disable import/prefer-default-export */
import { initState } from './simpleObject.middleware'
import { getState } from '../../selectors'

export const get = options => () => state => getState(options)(state)

export const isInitialized = options => (state) => {
  if (options.defaultData !== undefined) return state[options.name] !== options.defaultData
  return get(options)()(state) !== initState
}
