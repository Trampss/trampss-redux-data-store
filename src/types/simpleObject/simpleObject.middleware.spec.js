/* eslint-env jest */
import { set, reset, update } from '../../actions'
import simpleObject from './simpleObject.middleware'

const prefix = 'testPrefix'
const name = 'testName'
const Element = code => ({ code, some: 'other', infos: code })
const state = Element('elm')

describe('middlewares/simpleObject', () => {
  // Use the factory to create a new reducer named 'testPrefix'
  // It uses 'code' as key in givent elements
  const testPrefix = simpleObject('code')(prefix)(name)

  it('should initialize', () => expect(testPrefix(/* defaultData */)()).toMatchSnapshot())

  it('should initialize with defaultData', () => expect(testPrefix('default')()).toMatchSnapshot())

  it('should set element [elm2]', () => expect(
    testPrefix(/* defaultData */)({
      state,
      action: set(prefix)(name)(Element('elm2')),
    }),
  ).toMatchSnapshot())

  it('should reset state', () => expect(
    testPrefix(/* defaultData */)({
      state,
      action: reset(prefix)(name)(),
    }),
  ).toMatchSnapshot())

  it('should update element', () => expect(
    testPrefix(/* defaultData */)({
      state,
      action: update(prefix)(name)({ some: 'other 2', modifi: 'cation' }),
    }),
  ).toMatchSnapshot())

  it('should take defaultData param to populate data field -init-', () => {
    // init - with empty object
    expect(testPrefix({})()).toMatchSnapshot()
    // init - with object
    expect(testPrefix({ im: 'default' })()).toMatchSnapshot()
    // init - with empty string
    expect(testPrefix('')()).toMatchSnapshot()
  })
  it('should take defaultData param to populate data field -reset-', () => {
    // reset
    expect(testPrefix({ im: 'default' })({ state, action: reset(prefix)(name)() })).toMatchSnapshot()
  })
})
