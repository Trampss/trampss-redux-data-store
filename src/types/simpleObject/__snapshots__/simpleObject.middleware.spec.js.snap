// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middlewares/simpleObject should initialize 1`] = `
Object {
  "state": Object {},
}
`;

exports[`middlewares/simpleObject should initialize with defaultData 1`] = `
Object {
  "state": "default",
}
`;

exports[`middlewares/simpleObject should reset state 1`] = `
Object {
  "action": Object {
    "type": "@@krf/RESET_TESTPREFIX",
  },
  "state": Object {},
}
`;

exports[`middlewares/simpleObject should set element [elm2] 1`] = `
Object {
  "action": Object {
    "payload": Object {
      "code": "elm2",
      "infos": "elm2",
      "some": "other",
    },
    "type": "@@krf/SET_TESTPREFIX",
  },
  "state": Object {
    "code": "elm2",
    "infos": "elm2",
    "some": "other",
  },
}
`;

exports[`middlewares/simpleObject should take defaultData param to populate data field -init- 1`] = `
Object {
  "state": Object {},
}
`;

exports[`middlewares/simpleObject should take defaultData param to populate data field -init- 2`] = `
Object {
  "state": Object {
    "im": "default",
  },
}
`;

exports[`middlewares/simpleObject should take defaultData param to populate data field -init- 3`] = `
Object {
  "state": "",
}
`;

exports[`middlewares/simpleObject should take defaultData param to populate data field -reset- 1`] = `
Object {
  "action": Object {
    "type": "@@krf/RESET_TESTPREFIX",
  },
  "state": Object {
    "im": "default",
  },
}
`;

exports[`middlewares/simpleObject should update element 1`] = `
Object {
  "action": Object {
    "payload": Object {
      "modifi": "cation",
      "some": "other 2",
    },
    "type": "@@krf/UPDATE_TESTPREFIX",
  },
  "state": Object {
    "code": "elm",
    "infos": "elm",
    "modifi": "cation",
    "some": "other 2",
  },
}
`;
